src/conventions/h-group/clue-interpretation/interpret-clue.js

```diff
@@ -146,6 +146,73 @@ function resolve_clue(game, old_game, action, inf_possibilities, focused_card) {
 	reset_superpositions(game);
 }
 
+/**
+ * Finalizes the bluff connections.
+ * @param {Game} game
+ * @param {number} giver
+ * @param {number} target
+ * @param {FocusPossibility[]} connections
+ * @returns {FocusPossibility[]}
+ */
+export function finalize_bluff_connections(game, giver, target, connections) {
+	const { state } = game;
+
+	const bluff_seat = (giver + 1) % state.numPlayers;
+	const no_bluff_connections = connections.some(conn =>
+		conn.connections.length > 0 && (
+			// If there's a visible connection outside of the bluff seat, expect them to play.
+			conn.connections[0].reacting !== bluff_seat && (bluff_seat == state.ourPlayerIndex || target !== state.ourPlayerIndex) ||
+			// If there's a non-bluff non-finesse connection following this play, it could be accidentally played making this an invalid bluff
+			conn.connections[0].reacting == bluff_seat && !conn.connections[0].bluff && (conn.connections.length == 1 || conn.connections[1].reacting !== bluff_seat && conn.connections[1].type != 'finesse')));
```

Comment:
Could you explain this block a bit more? I've been looking at it for a while and I'm not really sure what it's doing, or why it should be done this way. :sweat_smile:

Reply:
The first line is attempting to recognize cases where there is an interpretation where someone else plays first,
The second line is trying to detect when there is an interpretation where the same bluff seat player plays, but not as a bluff, and thus others can conclude that given their card connects to the bluff play they can play. My thinking is that this would detect these cases where it's a different line of reasoning that leads to the play after the bluff card.

To break it down, consider the negation of this:
* If the first play of the connection is not the bluff seat, then playing the bluffed card would reveals that the other connection is not the real one,
* If it's a bluff connection, then false subsequent connections have been ruled out by resolve_bluff. 
* If there is only a single connection, then there is an interpretation where this play directly leads to the focused card as a non-bluff play, so playing an alternate bluff interpretation would be ambiguous.
* if the next connection is another player, then similar to the previous there is a non-bluff interpretation where some other player plays after the bluff card play.

Note that, in explaining this i realize i should be checking that the first play is indeed the finesse as if it's playing some other card then it wouldn't be ambiguous.

Reply:
> Note that, in explaining this i realize i should be checking that the first play is indeed the finesse as if it's playing some other card then it wouldn't be ambiguous.

Of course if there exists an interpretation where the first play is not a blind play then a bluff shouldn't be considered anyways.

Reply:
I think I'm starting to get it. It's a bit like *Cathy's Connecting Principle* but after we've determined all the potential connections it could be, right? But why was the old condition (a connection exists not starting in bluff seat) insufficient?
- If we are the clue target and see a connection not starting on us, why should we still consider a bluff?
- If the player in bluff seat plays a card not as a bluff, why does it matter who the next connecting player is or what their connection is? Wouldn't it be impossible for a bluff to exist, since it would become delayed?

Maybe an example of the two cases and how the old condition didn't catch them correctly would help, if it's not too difficult to construct.

Reply:
If this is necessary to remove the p3 interpretation, I think that should actually be removed at the `known_bluff` level (line 224 of connecting-cards.js), because the interpretation `[4 t2 bluff (will-bot2) -> 2 p2 known (will-bot2) -> p3?]` shouldn't be valid. Blind playing a 2 connects to a 3 clue, so this can't be a bluff. Instead of comparing with the next promised card, shouldn't we be checking the clued card and also whether the clue matches? e.g.

```js
const known_bluff = bluff_identities.some(i =>
    (clue.type === CLUE.RANK ? i.rank !== clue.value - 1 : !cardTouched(i, state.variant, clue)) ||          // card doesn't connect to clue
    !game.common.thoughts[focusedCard.order].inferred.has({ suitIndex: i.suitIndex, rank: i.rank + 1 }));    // impossible to connect via empathy
```

Modifying `resolve_bluff()` accordingly to pass in the ClueAction passes all tests with the old `no_bluff_connections` condition.

Reply:
> Maybe an example of the two cases and how the old condition didn't catch them correctly would help, if they're not too difficult to construct.

The main examples are the ones in the resolved issue and in the added test case, but they aren't complicated enough to demonstrate the problem with trying to determine it in resolve_bluff.

> If this is necessary to remove the p3 interpretation in #265, I think that should actually be removed at the `known_bluff` level (line 224 of connecting-cards.js), because the interpretation `[4 t2 bluff (will-bot2) -> 2 p2 known (will-bot2) -> p3?]` shouldn't be valid. Blind playing a 2 connects to a 3 clue, so this can't be a bluff. Instead of comparing with the next promised card, shouldn't we be checking the clued card and also whether the clue matches? e.g.
> 
> ```js
> const known_bluff = bluff_identities.some(i =>
>     (clue.type === CLUE.RANK ? i.rank !== clue.value - 1 : !cardTouched(i, state.variant, clue)) ||          // card doesn't connect to clue
>     !game.common.thoughts[focusedCard.order].inferred.has({ suitIndex: i.suitIndex, rank: i.rank + 1 }));    // impossible to connect via empathy
> ```
> 
> Modifying `resolve_bluff()` accordingly to pass in the ClueAction passes all tests with the old `no_bluff_connections` condition.

This works in the case where the focused card would be the next play after the potentially bluffed card, but what if it's not? E.g. What if the focused card is the p4 in someone else's hand and the p3 is a prompt.

The problem with trying to fix this in resolve_bluff is that it's only looking at the bluff connection itself. We need to know that nobody will think they should play after the bluffed card plays.

Reply:
From my understanding of [*Cathy's Connecting Principle*](https://hanabi.github.io/level-11#cathys-connecting-principle-part-2), everyone can use this "suit/rank connect" matching to tell if it's a bluff or not. Everyone knows what clue was given and what card blind plays (except for the blind player, but Bob always assumes the truth), so as long as we can tell whether it connects or not, everyone should be able to tell.

> E.g. What if the focused card is the p4 in someone else's hand and the p3 is a prompt. 

If you mean something like: (p1 on stacks)
Alice: g1 xx xx xx
Bob: xx xx p[3] xx
Cathy: xx p4 xx xx
Donald: xx xx xx xx
Then if Donald clues 4 to Cathy, Alice will play g1 and 1 doesn't connect to 4 (i.e. the clue value is not 1 greater than the blind played card's rank, `known_bluff` is true), so this bluff is valid and everyone knows.

Otherwise, something like: (g2, p1 on stacks)
Alice: g3 xx xx xx
Bob: xx xx p[3] xx
Cathy: xx p4 xx xx
Donald: xx xx xx xx
Then if Donald clues 4 to Cathy, Alice will play g3 and 3 connects to 4 (i.e. the clue value is 1 greater than the blind played card's rank, `known_bluff` is false), so this bluff is invalid because the clued card will look like g4.

I think I'm not understanding the problem you're trying to describe, could you elaborate on "an alternate connection possibility is a play in a different connection line from the bluff connection line"?

Reply:
> (i.e. the clue value is not 1 greater than the blind played card's rank, known_bluff is true)

This isn't required when [bluffing through already clued cards](https://hanabi.github.io/level-11#bluffs-through-already-clued-cards). Let me see if i can concoct a situation where this would come up:

Alice: b2 xx xx xx
Bob: r4 [b]5 xx xx
Cathy: xx xx xx xx
Donald: xx xx xx xx

b1, r2 is on the stack

Donald clues 4 to Bob.
Alice plays b1
The bluff connection b1 -> r4 obviously doesn't connect so Bob knows not to play the 4 immediately, but Bob can also see the possibility of b2 -> b3 (prompt) -> b4 as another connection and might think to play b3.

If Bob assumes the truth, Bob bombs the b5 (thinking it's a b3), but given the b5 wasn't part of the bluff "connection" how would Donald know it's not a valid bluff?

Reply:
I guess if you feel pretty sure that only checking the next play and focus in the bluff connection is sufficient we could try it and worst case revisit if/when situations like this come up, but it didn't sit right with me that the current code missed this because the "next play" depends on which connection you're interpreting it as.

Reply:
Your example looks like a valid bluff to me. The player Bob is in Cathy's seat, so he uses *Cathy's Connecting Principle* to tell whether a bluff or finesse occurred. From the docs, "if it is ambiguous whether or not the card connects, then everyone should assume a *Bluff* over a *Finesse*". Only the blind player uses *Bob's Truth Principle*: in this case Alice, who believed she was playing r3 on her turn.

This is an example of a [*Hard Bluff*](https://hanabi.github.io/level-13#the-hard-bluff), which is formally introduced at level 13. In fact, your situation is almost the same as the 2nd example in the docs. Since r4 is a valid bluff target, Bob (and everyone else) assumes that Alice played b2 into a bluff and no one is promised to have b3, although Bob could certainly have b4 and b3 as a bluff through already-clued cards.

> the current code missed this because the "next play" depends on which connection you're interpreting it as

Right, I think I understand what you were referring to. It doesn't really matter what the next play is (the old code was checking against `promised[1]`), because like you mentioned that can change depending on what the identity of the focused card is, and the receiver doesn't know what that is. The blind played card should be compared against the clue: everyone except the blind player can see the blind played card, and the blind player assumes the truth until the card is played, and afterward everyone can see it. There's just a bit of an extra check if the empathy on the clued card can't possibly connect.

Reply:
Thanks for the reference to hard bluffs. Those rules make things *much* simpler. If that's true I think all of that complex logic about determining whether the next play looks playable can be replaced with a much simpler check of whether the blind play matches the focus, with an assumption that it does if the player doesn't know the identity of the blind play (i.e. the person playing into the bluff).